MaxText/decode_multi.py

```diff
@@ -0,0 +1,218 @@
+        true_length=true_length,
+        rng=rng_prefill,
+        slot=slot_idx,
+        request_id=request_id,
+    )
+    prefill_results_to_insert[slot_idx] = prefill_result
+    streams_results[slot_idx].append(first_token.get_result_at_slot(0).tokens.item())
+    streams_prefilled_count += 1
+    print(f"After prefill stream {slot_idx}")
+
+  # --- Insert Initial Prefills ---
+  print("Inserting initial prefill results...")
+  for slot_idx, prefill_result in prefill_results_to_insert.items():
+    request_id = uuid.uuid4()
+    decode_state = engine.insert(
+        prefix=prefill_result,
+        decode_state=decode_state,
+        slot=slot_idx,
+        request_id=request_id,  # Pass request_id
+    )
+    streams_active[slot_idx] = True  # Mark stream as active
+    streams_inserted_count += 1
+    print(f"  Inserted prefill for slot {slot_idx}")
+    max_utils.print_mem_stats(f"After inserting stream {slot_idx}")
+
+  print("Starting interleaved generation loop...")
+  total_steps = config.max_target_length - config.max_prefill_predict_length
+  for step in range(total_steps):
+    print(f"\n--- Step {step + 1} / {total_steps} ---")
+
+    # Generate step for all active streams
+    active_stream_indices = [i for i, active in enumerate(streams_active) if active and not streams_finished[i]]
+    if active_stream_indices:
+      print(f"  Generating for active slots: {active_stream_indices}")
+      rng, rng_generate = jax.random.split(rng)
+      decode_state, sampled_tokens = engine.generate(params, decode_state, rng=rng_generate)
+      max_utils.print_mem_stats(f"  After generate step {step+1}")
+
+      # Store the generated token and check for finished streams
+      for slot_idx in active_stream_indices:
+        # Check if the stream finished this step
+        current_len = config.max_prefill_predict_length + step + 1  # Includes prefill + current step
+        finished_this_step = False
+        if current_len >= config.max_target_length:
+          print(f"    Stream in slot {slot_idx} reached max target length.")
+          streams_finished[slot_idx] = True
+          streams_active[slot_idx] = False
+          finished_this_step = True
+
+        # Store token if it wasn't already finished before this step or if it finished on this step
+        if not streams_finished[slot_idx] or finished_this_step:
+          # Ensure we don't try to access results for a slot that might not exist
+          if slot_idx < sampled_tokens.data.shape[0]:
+            token_for_slot = sampled_tokens.get_result_at_slot(slot_idx).tokens.item()
+            streams_results[slot_idx].append(token_for_slot)
+          else:
+            print(f"Warning: Tried to get token for slot {slot_idx}, but batch size seems smaller.")
+
+        # Call release_pages if finished this step
+        if finished_this_step:
+          print(f"    Calling engine to release pages for finished slot {slot_idx}...")
+          engine.release_pages(slot=slot_idx)
```

Comment:
qq, why do we need to manually release the pages. My understanding is: In prefill stage, when we allocate this slot to for the request, we will firstly release the pages for this slot. So we don't need to worry about clean up here?

Reply:
1. This is just a demo script showing how it _can_ be used so it's not required to be used this way. 
2. There are still valid cases for manually releasing pages via the engine api. I.e. to return pages to the pool if there is an error at the orchestration level without needing to wait for a prefill to be called.